### PR TITLE
Small mappers update.

### PR DESCRIPTION
**Pull request by Zemurin**
Just a few mappers to set up skeleton for EU4World.

### CODE REVIEW

File: CK3ToEU4Tests/MapperTests/CultureMapper/CultureMappingTests.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
+#include "../../CK3toEU4/Source/Mappers/CultureMapper/CultureMappingRule.h"
```
Comment (Idhrendur) [inline CK3ToEU4Tests/MapperTests/CultureMapper/CultureMappingTests.cpp:1]: Make the tests filename match the source file (add 'rule')

File: CK3ToEU4Tests/MapperTests/CultureMapper/CultureMappingTests.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+#include "gtest/gtest.h"
+#include <sstream>
+
+TEST(Mappers_CultureMappingTests, eu4CultureDefaultsToBlank)
```
Comment (Idhrendur) [inline CK3ToEU4Tests/MapperTests/CultureMapper/CultureMappingTests.cpp:5]: Same with the tests class

Comment (Idhrendur) [inline CK3ToEU4Tests/MapperTests/CultureMapper/CultureMappingTests.cpp:5]: Also, ideally, you should devise tests that exercise the intended behavior of the class, not check its internal state (especially when that involves opening a bunch of getters you don't really care about in the end).

File: CK3toEU4/Source/Mappers/RegionMapper/RegionMapper.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#include <map>
+
+class Configuration;
+namespace mappers
+{
+class RegionMapper: commonItems::parser
+{
+  public:
+	RegionMapper() = default;
+	virtual ~RegionMapper() = default;
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/RegionMapper/RegionMapper.h:17]: Why the virtual destructor. Are you inheriting from this class (including mocks)? If not, delete the virtual.

At which point you can just delete all the default big six.

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+namespace CK3
+{
+class Title;
+}
+
+namespace mappers
+{
+class ProvinceMapper: commonItems::parser
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.h:16]: Should expand the tests to cover the newer functions like looking up using county name or looking up titles.

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+		for (auto ck3ID: mapping->second)
+		{
+			if (baroniesToCounties.count(ck3ID))
+			{
+				auto targetTitle = baroniesToCounties.find(ck3ID)->second;
+				toReturn.insert(targetTitle);
+			}
+			else
+			{
+				throw std::runtime_error(
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:85]: You should probably also devise a test for this edge case, too.

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+					 "Province mapper error - requested ck3 province ID " + std::to_string(ck3ID) + " is invalid. We cannot find related county. Farewell.");
+			}
+		}
+		return toReturn;
+	}
+	else
+	{
+		throw std::runtime_error("Province mapper error - requested ck3 province ID does not exist in mappings.");
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:93]: And this. If you're going to die, document it with a test.

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	if (mapping != countiesToBaronies.end())
+	{
+		const auto& remapping = CK3ToEU4ProvinceMap.find(mapping->second);
+		if (remapping != CK3ToEU4ProvinceMap.end())
+		{
+			return remapping->second;
+		}
+		else
+		{
+			Log(LogLevel::Warning) << "Province Mapper mismatch - " << ck3CountyName << " with baronyID " << mapping->second << " has no mappings!";
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:109]: ditto

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		}
+	}
+	else
+	{
+		Log(LogLevel::Warning) << "Province Mapper mismatch - " << ck3CountyName << " has no mappings!";
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:114]: ditto

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+			{
+				// This is a sane barony and we know of its troubles.
+				baroniesToCounties.insert(std::pair(ck3ID, theMatrix[ck3ID]));
+				countiesToBaronies.insert(std::pair(theMatrix[ck3ID].first, ck3ID));
+				++counter;
+			}
+			else
+			{
+				// We're mapping to something unheard of. Let's not.
+				throw std::runtime_error("Province mapping ck3 = " + std::to_string(ck3ID) + " is invalid. We cannot find related county. Farewell.");
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:158]: ditto

File: CK3ToEU4Tests/MapperTests/RulerPersonalityMapper/RulerPersonalitiesMappingTests.cpp [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+{
+	std::stringstream input;
+	input << "= {}";
+
+	const mappers::RulerPersonalitiesMapping theMapping(input);
+
+	ASSERT_TRUE(theMapping.getTraits().empty());
+}
+
+TEST(Mappers_RulerPersonalitiesMappingTests, traitsCanBeLoaded)
```
Comment (Idhrendur) [inline CK3ToEU4Tests/MapperTests/RulerPersonalityMapper/RulerPersonalitiesMappingTests.cpp:15]: I'd drop this test and the corresponding function. The other tests prove what you want.

File: CK3toEU4/Source/Mappers/RulerPersonalitiesMapper/RulerPersonalitiesMapper.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace mappers
+{
+class RulerPersonalitiesMapper: commonItems::parser
+{
+  public:
+	RulerPersonalitiesMapper();
+	explicit RulerPersonalitiesMapper(std::istream& theStream);
+
+	[[nodiscard]] std::set<std::string> evaluatePersonalities(const std::pair<int, std::shared_ptr<CK3::Character>>& theCharacter) const;
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/RulerPersonalitiesMapper/RulerPersonalitiesMapper.h:21]: You only use the latter part of that pair, so just take the character as input. Better yet, you only use the traits from the character, so take just the traits (unless that pushes too much abstraction up to the callers).

File: CK3ToEU4Tests/MapperTests/TitleTagMapper/TitleTagMapperTests.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+{
+	std::stringstream input;
+
+	const mappers::TitleTagMapper theMapper(input);
+
+	ASSERT_TRUE(theMapper.getMappings().empty());
+	ASSERT_TRUE(theMapper.getRegisteredTitleTags().empty());
+}
+
+TEST(Mappers_TitleTagMapperTests, canLoadTitleTagLinks)
```
Comment (Idhrendur) [inline CK3ToEU4Tests/MapperTests/TitleTagMapper/TitleTagMapperTests.cpp:15]: You can drop this test and the associated function

File: CK3toEU4/Source/Mappers/TitleTagMapper/TitleTagMapper.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+std::optional<std::string> mappers::TitleTagMapper::getTagForTitle(const std::string& ck3Title, int eu4Capital)
+{
+	// the only case where we fail is on invalid invocation. Otherwise, failure is
+	// not an option!
+	if (ck3Title.empty())
+		return std::nullopt;
+
+	// The pope don't use proper title and isn't registered.
+	if (ck3Title == "The Pope")
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/TitleTagMapper/TitleTagMapper.cpp:49]: Add tests for this.

File: CK3toEU4/Source/Mappers/TitleTagMapper/TitleTagMapping.h [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+{
+  public:
+	TitleTagMapping() = default;
+	explicit TitleTagMapping(std::istream& theStream);
+
+	[[nodiscard]] std::optional<std::string> titleMatch(const std::string& ck3title) const;
+	[[nodiscard]] std::optional<std::string> capitalMatch(int eu4capital) const;
+	[[nodiscard]] bool fallbackMatch() const;
+
+	[[nodiscard]] const auto& getEU4Tag() const { return eu4Tag; }		 // used for testing
```
Comment (Idhrendur) [inline CK3toEU4/Source/Mappers/TitleTagMapper/TitleTagMapping.h:19]: You should be able to drop these and the associated tests. You're testing the functionality you actually care about.

**Review by Idhrendur — APPROVED** [review]

File: CK3toEU4/Source/Mappers/RulerPersonalitiesMapper/RulerPersonalitiesMapper.h (hunk 11 above)
Comment (Zemurin) [reply CK3toEU4/Source/Mappers/RulerPersonalitiesMapper/RulerPersonalitiesMapper.h:21]: I know and would agree, but I find that shoving entire pairs is much cleaner in the functions that do actual stuff. I treat characters, titles, provinces and everything else as pairs and in 95% cases I need both elements. 

So when I say "I need personalities of this character" then the syntax .evaluatePersonalities(character) simply looks better than .evaluatePersonalities(character->second->getTraits());

Noone looks in the mappers anyway once they pass tests, while the code upstairs gets reshuffled quite often.

File: CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp (hunk 7 above)
Comment (Zemurin) [reply CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp:109]: This is not a fail, just a warning. I'm expecting potential mis-lookups from CK3 custom titles and will adjust this to exception or remove warning entirely depending on live data tests.
